Modules/StudyProgramme/classes/class.ilObjStudyProgrammeSettingsGUI.php

```diff
@@ -191,50 +213,79 @@ protected function buildModalHeading($label, $current_node) {
+						self::PROP_DESC =>
+							$ff->text($txt("description")) // TODO: change this to textarea once available
+								->withValue($prg->getDescription())
+					],
+					$txt("prg_edit"),
+					""
+				),
+				$ff->section(
+					[
+						self::PROP_TYPE =>
+							$ff->select($txt("type"), $sp_types)
+								->withValue((string)$prg->getSubtypeId())
+								->withAdditionalTransformation($tf->custom(function($v) {
+									if ($v == "") {
+										return 0;
+									}
+									return $v;
+								}))
+					],
+					$txt("prg_type"),
+					""
+				),
+				$ff->section(
+					[
+						self::PROP_POINTS =>
+							$ff->numeric($txt("prg_points"))
+								->withValue((string)$prg->getPoints()),
+						self::PROP_STATUS =>
+							$ff->select($txt("prg_status"), $status_options)
+								->withValue((string)$prg->getStatus())
+					],
+					$txt("prg_assessment"),
+					""
+				)
+			]
+		)
+		->withAdditionalTransformation($tf->custom(function($values) {
+			// values now contains the results of the single sections,
+			// i.e. a list of arrays that each contains keys according
+			// to the section they originated from.
+			return call_user_func_array("array_merge", $values);
+		}));
+	}
 
-		if(!$this->ctrl->isAsynch()) {
-			$form->setAsync(false);
+	protected function updateWith(\ilObjStudyProgramme $prg, array $data) {
+		$prg->setTitle($data[self::PROP_TITLE]);
+		$prg->setDescription($data[self::PROP_DESC]);
+
+		if($prg->getSubtypeId() != $data[self::PROP_TYPE]) {
+			$prg->setSubtypeId($data[self::PROP_TYPE]);
+			$prg->updateCustomIcon();
+			$this->parent_gui->setTitleAndDescription();
 		}
 
-		$form->setFormAction($this->ctrl->getFormAction($this));
-		
-		$header = new ilFormSectionHeaderGUI();
-		$header->setTitle($this->lng->txt("prg_edit"));
-		$form->addItem($header);
-		
-		$item = new ilTextInputGUI($this->lng->txt("title"), self::PROP_TITLE);
-		$item->setRequired(true);
-		$form->addItem($item);
-		
-		$item = new ilTextAreaInputGUI($this->lng->txt("description"), self::PROP_DESC);
-		$form->addItem($item);
-		
-		$header = new ilFormSectionHeaderGUI();
-		$header->setTitle($this->lng->txt("prg_type"));
-		$form->addItem($header);
-		
-		$item = new ilSelectInputGUI($this->lng->txt("type"), self::PROP_TYPE);
-		$item->setOptions(ilStudyProgrammeType::getAllTypesArray());
-		$form->addItem($item);
-		
-		$header = new ilFormSectionHeaderGUI();
-		$header->setTitle($this->lng->txt("prg_assessment"));
-		$form->addItem($header);
-		
-		$item = new ilNumberInputGUI($this->lng->txt("prg_points"), self::PROP_POINTS);
-		$item->setMinValue(0);
-		$form->addItem($item);
-		
-		$item = new ilSelectInputGUI($this->lng->txt("prg_status"), self::PROP_STATUS);
-		$item->setOptions(self::getStatusOptions());
-		$form->addItem($item);
-		
-		$form->addCommandButton("update", $this->lng->txt("save"));
-		$form->addCommandButton("cancel", $this->lng->txt("cancel"));
-		
-		return $form;
+		$prg->setPoints($data[self::PROP_POINTS]);
```

Comment:
What about input filtering (at least with known mechanisms like `ilUtil::stripSlashes()` implicitly done in legacy form handling)? Currently the raw HTTP post data (checked with constraints) is persisted in database by calling `$prg->update();`. I don't see any sanitizing step here. When merging this, this process will be vulnerable against *XSS*.

Reply:
Although `$data` might look superficially similar to raw HTTP-post data, it in fact is already processed and "filtered". The core of this functionality is [here](https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/src/UI/Implementation/Component/Input/Field/Input.php#L354). The values supplied via post are checked with `isClientSideValueOk` first. This is `is_numeric` for the Numeric Field. If this does not work, an exception is thrown. The values are then checked and transformed via additional validations and transformations, where Numeric Field again checks for is_numeric via a constraint (which is already to much, because the value certainly is numeric already) Only if this worked for every field in the form, the user gets the checked and transformed data from [`Form::getData`](https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/src/UI/Implementation/Component/Input/Container/Form/Form.php). There might be bugs in this mechanism (as always) and there are fields (I'm looking at you, `Select Field`) that are missing necessary checks, but this is in now way raw data from HTTP-Post. Also I cannot see why we would want to treat every input with `ilUtil::stripSlashes` if there are simpler means (like checking `is_numeric` for the `Numeric Field`, or checking against available options for `Select Field`) to discard malicious input. We might need to discuss how to treat general text-input in that regard, though.

Reply:
Maybe my comment points to the wrong line of code. It is not about the numeric fields (here: points). It's about the title and description. I could not see any constraints bound to these fields. And of course something like `ilUtil::stripSlashes()` MUST NOT be part of the `Input` service, but I expected something like a `MustNotContainInvalidChars` constraint or some kind of transformation (making use of `\ilUtil::stripSlashes()`, or the `Input filter` service in future) for the text fields already shipped with this initial/first form migration.